internal/k8s/replicaset/replicaset.go

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package replicaset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+)
+
+// ReplicaSetWatcher is a type alias for k8s resource controller.
+type ReplicaSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx         context.Context
+	mgr         manager.Manager
+	name        string
+	namespace   string
+	onError     func(err error)
+	onReconcile func(rs map[string][]ReplicaSet)
+}
+
+// ReplicaSet is a type alias for the k8s replica set definition.
+type ReplicaSet = appsv1.ReplicaSet
+
+// New returns the ReplicaSetWatcher that implements reconciliation loop, or any error occurred.
+func New(opts ...Option) (ReplicaSetWatcher, error) {
+	r := new(reconciler)
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrieve the ReplicaSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	rsl := new(appsv1.ReplicaSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, rsl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Error("not found", err)
+			res.RequeueAfter = time.Second
+			return res, nil
+		}
+		return
+	}
+
+	rs := make(map[string][]ReplicaSet, len(rsl.Items))
+
+	for _, replicaset := range rsl.Items {
+		name, ok := replicaset.GetObjectMeta().GetLabels()["app"]
+		if !ok {
+			pns := strings.Split(replicaset.GetName(), "-")
+			name = strings.Join(pns[:len(pns)-1], "-")
+		}
+
+		if _, ok := rs[name]; !ok {
+			rs[name] = make([]ReplicaSet, 0, len(rsl.Items))
+		}
+
+		rs[name] = append(rs[name], replicaset)
+	}
```

Comment:
i think it is not so meaningful to put them into a map here.
Do you have a vision to use them?
i think it is enough to use just a slice.
```go
	onReconcile func(rs []ReplicaSet)
```

Reply:
@rinx 
If we use `[]Replicaset`, we have to loop each app for getting Metadata.
On the other hand, when we use `map[string][]Replicaset`, we can access easily each app's Metadata and do not have to loop for getting it.

Reply:
> If we use []Replicaset, we have to loop each app for getting Metadata.

already you have a loop to get `metadata.labels` in L88-L100 even if you use `map[string][]ReplicaSet` type for the argument of `onReconcile`.

even if you use `map[string][]ReplicaSet` type, i think you may have a loop for `[]ReplicaSet` in the `onReconcile` function like:
```go
func onReconcile(rs map[string][]ReplicaSet) {
    xrs, ok := rs["hoge"]
    // ... omission ...
    for _, r := range xrs {
        // do something
    }
}
```
so the loop for `[]ReplicaSet` may duplicate.

This is just a comment. so you don't have to revise it, but i would write it without map.

Reply:
thx, and we just decided use this style.
But, we should consider the performance later, we will consider about it again when simulating performance.

Reply:
As you know the k8s/pod,node package uses the same logic, but since the Reconciler can retrieve resources from the entire namespace, you'll need to turn the long loop and do a string comparison of if == when retrieving a Resource with a specific name.
Therefore, by using the fact that the time axis of Reconcile and the time axis for processing Cache are different, it is possible to process only a list of specific resource names in the onReconcile function by grouping resources into a map for each resource name at Reconcile.